cnudie/upload.py

```diff
@@ -0,0 +1,135 @@
+import enum
+import hashlib
+
+import ccc.oci
+import gci.componentmodel as cm
+import gci.oci
+import json
+import oci.model as om
+import dataclasses
+
+
+class UploadMode(enum.Enum):
+    SKIP = 'skip'
+    FAIL = 'fail'
+    OVERWRITE = 'overwrite'
+
+
+def _normalise_component_name(component_name: str) -> str:
+    return component_name.lower()  # oci-spec allows only lowercase
+
+
+def _target_oci_ref(
+    component: cm.Component,
+    component_ref: cm.ComponentReference=None,
+    component_version: str=None,
+):
```

Comment:
I think this signature should be overhauled (I know this function was copied; still I think we should take this opportunity and improve it).

suggestion (pseudo-code-ish):

```
def _oci_artefact_reference(
  component: Component | ComponentIdentity | ComponentReference | str # if str, expect `name:version` | tuple[str, str] # name, version),
  ocm_repository: str | RepoDataClass # did not lookup exact name, see gci.model = None,
) -> str: # is it actually str?
  # if component, ocm-repo can be omitted (but must be honoured if given)
  # otherwise, ocm-repo must be passed
```

Reply:
Done